src/raqm.c

```diff
@@ -1123,38 +1214,80 @@ _raqm_itemize (raqm_t *rq)
   }
 
   if (rq->base_dir == RAQM_DIRECTION_RTL)
-    par_type = FRIBIDI_PAR_RTL;
+    par_type = RAQM_BIDI_PAR_RTL;
   else if (rq->base_dir == RAQM_DIRECTION_LTR)
-    par_type = FRIBIDI_PAR_LTR;
+    par_type = RAQM_BIDI_PAR_LTR;
 
   if (rq->base_dir == RAQM_DIRECTION_TTB)
   {
     /* Treat every thing as LTR in vertical text */
     max_level = 1;
-    memset (types, FRIBIDI_TYPE_LTR, rq->text_len);
+    memset (types, RAQM_BIDI_TYPE_LTR, rq->text_len);
     memset (levels, 0, rq->text_len);
     rq->resolved_dir = RAQM_DIRECTION_LTR;
   }
   else
   {
+#ifdef RAQM_SHEENBIDI
+    SBCodepointSequence codepointSequence =
+    {
+      SBStringEncodingUTF32,
+      (void *) rq->text,
+      rq->text_len,
+    };
+
+    /* paragraph */
+    bidiAlgorithm =
+      SBAlgorithmCreate (&codepointSequence);
+
+    firstParagraph =
+      SBAlgorithmCreateParagraph (bidiAlgorithm,
+                                  0,
+                                  INT32_MAX,
+                                  SBLevelDefaultLTR);
+
+    SBUInteger paragraphLength =
+      SBParagraphGetLength (firstParagraph);
+
+    /* lines */
+    paragraphLine =
+      SBParagraphCreateLine (firstParagraph,
+                             0,
+                             paragraphLength);
+
+    run_count =
+      SBLineGetRunCount (paragraphLine);
+
+    par_type =
+      SBParagraphGetBaseLevel (firstParagraph);
+
+    free (levels);
+    levels =
+      SBParagraphGetLevelsPtr (firstParagraph);
+
+    free (types);
```

Comment:
`levels` and `types` were freed below, but you made it FriBiDi `ifdef`'d. Make the others unconditional and don’t free here.

Reply:
fixed in 05f5c65467713f48df8b93a6e8be7ef109ec64af